ansible/inventory.go

```diff
@@ -0,0 +1,108 @@
+/*
+Copyright 2022.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ansible
+
+import "gopkg.in/yaml.v3"
+
+// Host represents ansible host
+type Host struct {
+	name string
+	Vars map[string]interface{} `yaml:",inline"`
+}
+
+// MakeHost instatiates an Host object
+func MakeHost(name string) Host {
+	return Host{
+		name: name,
+		Vars: make(map[string]interface{}),
+	}
+}
+
+// Group represents ansible group
+type Group struct {
+	name     string
+	Vars     map[string]interface{} `yaml:",omitempty"`
+	Hosts    map[string]Host        `yaml:",omitempty"`
+	Children map[string]Group       `yaml:",omitempty"`
+}
+
+//nolint:unused
+func (group Group) addHost(name string) Host {
+	host := MakeHost(name)
+	group.Hosts[name] = host
+	return host
+}
+
+// MakeGroup instatiates an Group object
+func MakeGroup(name string) Group {
+	return Group{
+		name:     name,
+		Vars:     make(map[string]interface{}),
+		Hosts:    make(map[string]Host),
+		Children: make(map[string]Group),
+	}
+}
+
+func (group Group) addChild(newGroup Group) Group {
+	group.Children[group.name] = newGroup
+	return newGroup
+}
+
+func (group Group) fixNames() {
+	for hostName, host := range group.Hosts {
+		host.name = hostName
+		group.Hosts[hostName] = host
+	}
+	for childName, child := range group.Children {
+		child.name = childName
+		child.fixNames()
+		group.Children[childName] = child
+	}
+
+}
+
+// Inventory contains parsed inventory representation
+type Inventory struct {
+	Groups map[string]Group `yaml:",inline,flow"`
+}
+
+// MakeInventory instatiates an Inventory object
+func MakeInventory() Inventory {
+	return Inventory{
+		Groups: make(map[string]Group),
+	}
+}
+
+func (inv Inventory) addGroup(name string) Group {
+	group := MakeGroup(name)
+	inv.Groups[name] = group
+	return group
+}
+
+func (inv Inventory) fixNames() {
+	for groupName, group := range inv.Groups {
+		group.name = groupName
+		group.fixNames()
+		inv.Groups[groupName] = group
+	}
+}
+
+// MarshalYAML serializes the Inventory object
+func (inv *Inventory) MarshalYAML() ([]byte, error) {
+	inv.fixNames()
```

Comment:
That isn't required. I only added fixNames for unmarshalling (since name will not have been marshalled). If you don't need unmarshalling support then can remove all of the fixNames functions.